File: IdeaProject/КИТПО/kitpo_lr_4/src/main/scala/MyTree.scala

import java.io.Serializable

import scala.collection.mutable
import scala.collection.mutable.Stack
abstract class MyTree extends   ITreeBuilder with Iterable[Any] with  Serializable
{
  class Node extends  Serializable
  {
    var pLeft: Node = null
    var pRight:  Node = null
    var data: Any = null
  }
  var myStr: String = ""
  var countEl: Int = 0
  var root: Node = null
  var _type: String = null
  def iterator: Iterator[Any] = {
    return new MyTreeIterator() {
    }
  }

  class MyTreeIterator extends Iterator[Any] with Serializable
  {
    var curr: Node = root
    var sizeIt: Int = 0
    var myvalues = Stack[Node]()
    pushAll(curr)

    def pushAll(cur: Node): Unit = {
      var myCur: Node = cur
      while(myCur!=null){
        myvalues.push(myCur)
        myCur = myCur.pLeft
      }
    }

    override def hasNext: Boolean = {
      !myvalues.isEmpty
    }

    override def next(): Any = {
      if (!hasNext) {
        throw new NoSuchElementException()
      }
      var delEl: Node = myvalues.pop()
      pushAll(delEl.pRight)
      return delEl.data
    }
  }

  def push(data: Any): Unit =
  {
    var a: Node = new Node()
    a.data = data
    a.pLeft = null
    a.pRight = null
    if (root == null)
    {
      countEl += 1
      root = a
    } else {
      pushR(root, a)
    }
  }

  def pushR(root: Node, newNode: Node): Unit =
  {

    if (comapareTo(root.data, newNode.data) == 1) {
      if(root.pLeft == null) {
        root.pLeft = newNode
        countEl += 1
      }else pushR(root.pLeft, newNode)
    } else if(comapareTo(root.data, newNode.data) == -1) {
      if(root.pRight == null) {
        root.pRight = newNode
        countEl += 1
      }
      else pushR(root.pRight, newNode)
    }
  }

  def balanse(): MyTree ={
    var items = new Array[Any](countEl)
    var i: Int = 0
    for(obj <- this){
      items(i) = obj
      i += 1
    }


    var l: MyTree = Clone();
    _balanse(items, l);

    return l;
  }

  def _balanse(items: Array[Any], l: MyTree): Unit = {
    var i: Int = 0
    if (items.length != 0){
      l.push(items(items.length / 2))
      var leftSize: Int = items.length / 2
      var rightSize: Int = items.length - leftSize - 1
      var left = new Array[Any](leftSize)
      var right = new Array[Any](rightSize)
      left = CopyArray(items, 0, left, leftSize)
      right = CopyArray(items, leftSize + 1, right, rightSize)
      _balanse(left, l)
      _balanse(right, l)
    }
  }

  def CopyArray(arr: Array[Any], start: Int, newArr: Array[Any], size: Int): Array[Any] = {
    for(i <- start to (start + size - 1)){
      newArr(i - start) = arr(i)
    }
    return newArr
  }

  def myprint(): Unit ={
    myStr = ""
    levelOrderTraversalUsingRecursion(root)
  }

  def get_at_index(index: Int): Any = {
    var h: Node = root;
    var cur: Int = 0;
    for(obj <- this)
    {
      cur+=1;
      if(cur == index) return obj
    }
    return null
  }
  def levelOrderTraversalUsingRecursion(curNode: Node): Unit ={
    if(curNode != null){
      var height = getHeightOfTree(curNode)
      for (i <- 0 to height) {
        printAtGivenLevel(curNode, i + 1)
        myStr += "\n"
      }

    } else myStr =""
  }

  def getHeightOfTree(curNode: Node): Int ={
    var currNode: Node = curNode
    if(currNode ==null) return -1
    return Math.max(getHeightOfTree(currNode.pLeft), getHeightOfTree(currNode.pRight)) + 1
  }

  def printAtGivenLevel(curNode: Node, level: Int): Unit = {

    if(curNode == null){
      myStr += " _ "
    }
    else if (level == 1) {
      myStr +=curNode.data + " "
    }
    else {
      printAtGivenLevel(curNode.pLeft, level - 1)
      printAtGivenLevel(curNode.pRight, level - 1)
    }
  }

  def del(index: Int): Unit ={
    var i: Int = 0
    for (obj <- this) {
      if(i == index) del_from_data( root, obj)
      i += 1
    }
  }
  def del_from_data(node: Node, data: Any): Node ={
    if(node == null) return node
    else{
      if(comapareTo(node.data, data) == 1) {
        node.pLeft = del_from_data(node.pLeft, data)
      } else if(comapareTo(node.data, data) == -1) {
        node.pRight = del_from_data(node.pRight, data)
      } else {
        if(node.pLeft == null && node.pRight == null) {
          return null
        } else if(node.pLeft == null) {
          return  node.pRight
        } else if(node.pRight == null) {
          return  node.pLeft
        } else {

          node.data = minValue(node.pRight)
          node.pRight=del_from_data( node.pRight, minValue(node.pRight))
        }
      }
      return node
    }
  }

  def minValue(node: Node): Any ={

    if(node.pLeft != null) {
      return minValue(node.pLeft)
    }else return node.data
  }

}

File: IdeaProject/КИТПО/kitpo_lr_2/src/MyTree.java
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Stack;

public abstract class MyTree  implements  ITreeBuilder, Serializable {

    public String seeTree;
    public Node root;
    public String type;

    MyTree(String type){
        root = null;
        this.type = type;
        seeTree ="";
    }
    public void copy(MyTree cur){
        this.root = cur.root;
        this.type = cur.type;
        this.seeTree = cur.seeTree;
    }


    private class Node implements Serializable {

        Node pLeft;
        Node pRight;
        Object data;
        Node() {
            pLeft = null;
            pRight = null;
        }

        Node(Object data) {
            this.data = data;
            pLeft = null;
            pRight = null;
        }




    }



    public void del(int index) {
        MyIterator iter = (MyIterator) getIterator();
        for(int i =0; i < index - 1; i++){
            if(iter.hasNext()) {
                iter.next();
            }
        }
        if(iter.hasNext()) {
            delFromData(root,  iter.next().data);
        }
    }

    private Node delFromData(Node root, Object data) {

        if(root == null) return root;

        if(comapareTo(root.data, data) == 1) {
            root.pLeft = delFromData(root.pLeft, data);
        } else if(comapareTo(root.data, data) == -1) {
            root.pRight = (delFromData(root.pRight, data));
        } else {
            if(root.pLeft == null && root.pRight == null) {
                return null;
            } else if(root.pLeft == null) {
                return  root.pRight;
            } else if(root.pRight == null) {
                return  root.pLeft;
            } else {
                Object minValue = minValue(root.pRight);
                root.data = (minValue);
                root.pRight=(delFromData( root.pRight, minValue));
            }
        }

        return root;
    }

    private Object minValue(Node node) {

        if(node.pLeft != null) {
            return minValue(node.pLeft);
        }
        return node.data;
    }


    public Object get(int index) {
        MyIterator iter = (MyIterator) getIterator();
        for(int i =0; i < index - 1; i++){
            if(iter.hasNext()) {
                iter.next();
            }
        }
        if(iter.hasNext())return iter.next().data;
        else return null;
    }


    public void add(Object data) {
        if(root != null){
            addR(root, data);
        }else{
            root = new Node(data);
        }
    }

    private void addR(Node cur, Object data) {
        if(comapareTo(cur.data, data)==0) return;
        else if(comapareTo(cur.data, data) == 1){
            if(cur.pLeft == null) {
                cur.pLeft = new Node(data);

            }
            else addR(cur.pLeft, data);
        }else if(comapareTo(cur.data, data) == -1){
            if(cur.pRight == null){

                cur.pRight= new Node(data);
            }
            else addR(cur.pRight, data);
        }
    }


    public MyTree balance() {
        MyTreeFactory f = new MyTreeFactory ();
        MyTree result = f.createTree(type);
        MyIterator iter = (MyIterator) getIterator();
        ArrayList<Object> myvalues = new ArrayList<>();
        ;
        while(iter.hasNext()) {myvalues.add( iter.next().data);}


        balanceH(myvalues, result);

        return result;
    }
    private void balanceH(ArrayList<Object> myvalues,  MyTree result) {
        ArrayList<Object> myvaluesR = new ArrayList<>();
        ArrayList<Object> myvaluesL = new ArrayList<>();
        if (myvalues.size() != 0) {
            result.add(myvalues.get(myvalues.size() / 2));

            for (int i = 0; i < myvalues.size(); i++) {
                if (i < myvalues.size() / 2) {
                    myvaluesR.add(myvalues.get(i));
                } else if (i > myvalues.size() / 2) {
                    myvaluesL.add(myvalues.get(i));
                }
            }
            balanceH(myvaluesR, result);
            balanceH(myvaluesL, result);
        } else return;
    }

        public void print(){
        seeTree = "";
        levelOrderTraversalUsingRecursion(root);
    }

    public void levelOrderTraversalUsingRecursion(Node node) {
        if (node == null) {
            seeTree="";
            return;
        }

        int height = getHeightOfTree(node);


        for(int i = 0; i <= height; i++) {
            printAtGivenLevel(node, i + 1);

            seeTree= seeTree + "\n";

        }
    }

    public int getHeightOfTree(Node node) {
        if(node == null){
            return -1;
        }

        return Math.max(getHeightOfTree(node.pLeft), getHeightOfTree(node.pRight)) + 1;
    }

    public void printAtGivenLevel(Node node, int level) {
        if(node == null) {

            seeTree=seeTree + " _ ";
            return;
        }

        if(level == 1) {

            seeTree=seeTree + node.data + " ";
            return;
        }

        printAtGivenLevel(node.pLeft, level - 1);
        printAtGivenLevel(node.pRight, level - 1);
    }


    public Iterator getIterator() {
        return new MyIterator(root);
    }

    private class MyIterator implements Iterator{

        private Stack<Node> stack;

        public MyIterator(Node root){
            stack  =new Stack<>();
            pushAll(root);
        }

        private void pushAll(Node root){

            while (root!=null) {
                stack.push(root);
                root = root.pLeft;
            }
        }


        public boolean hasNext() {
            return !stack.isEmpty();
        }


        public Node next() {
            Node node = stack.pop();
            pushAll(node.pRight);
            return  node;
        }
    }
}